node/node.go

```diff
@@ -274,24 +277,49 @@ func (n *Node) run(ctx context.Context) (err error) {
 			select {
 			case <-agentDone:
 				return
-			case node := <-n.notifyNodeChange:
-				// If the server is sending us a ForceRenewal State, renew
-				if node.Certificate.Status.State == api.IssuanceStateRotate {
+			case nodeChanges := <-n.notifyNodeChange:
+				if nodeChanges.Node != nil {
+					// If the server is sending us a ForceRenewal State, renew
+					if nodeChanges.Node.Certificate.Status.State == api.IssuanceStateRotate {
+						renewCert()
+						continue
+					}
+					n.Lock()
+					// If we got a role change, renew
+					role := ca.WorkerRole
+					if nodeChanges.Node.Role == api.NodeRoleManager {
+						role = ca.ManagerRole
+					}
+					if n.role == role {
+						n.Unlock()
+						continue
+					}
+					n.Unlock()
 					renewCert()
-					continue
 				}
-				n.Lock()
-				// If we got a role change, renew
-				role := ca.WorkerRole
-				if node.Role == api.NodeRoleManager {
-					role = ca.ManagerRole
-				}
-				if n.role == role {
+
+				if nodeChanges.RootCert != nil {
+					n.Lock()
+					role := n.role
 					n.Unlock()
-					continue
+
+					if role == ca.ManagerRole || bytes.Equal(nodeChanges.RootCert, securityConfig.RootCA().Certs) {
```

Comment:
We would miss the immediate one, but the next session message that changes the role will also come with the desired root cert, and since it is different than the one currently in security config, it should update.  E.g. it should eventually get the root certificate update, although perhaps not immediately.